src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelRecordReader.java

```diff
@@ -0,0 +1,158 @@
+import org.apache.http.message.BasicNameValuePair;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Scanner;
+
+/**
+ * RecordReader implementation, which reads events in json format from MixPanel api.
+ */
+public class MixPanelRecordReader extends RecordReader<NullWritable, Text> {
+  private static final Gson gson = new GsonBuilder().create();
+  private String currentEvent;
+  private CloseableHttpClient httpClient;
+  private Scanner dataScanner;
+  private CloseableHttpResponse response;
+
+  private HttpClientContext createContext(MixPanelBatchSourceConfig config) throws MalformedURLException {
+    URL mixPanelUrl = new URL(config.getMixPanelUrl());
+    HttpHost targetHost = new HttpHost(mixPanelUrl.getHost(), mixPanelUrl.getPort(), mixPanelUrl.getProtocol());
+    AuthCache authCache = new BasicAuthCache();
+    authCache.put(targetHost, new BasicScheme());
+
+    CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+    // MixPanel api auth performed by unique ID as username and empty password. No security issues here
+    // when https used as transport
+    credentialsProvider.setCredentials(
+      new AuthScope(mixPanelUrl.getHost(), mixPanelUrl.getPort()),
+      new UsernamePasswordCredentials(config.getApiSecret(), ""));
+
+    HttpClientContext context = HttpClientContext.create();
+    context.setCredentialsProvider(credentialsProvider);
+    context.setAuthCache(authCache);
+    return context;
+  }
+
+  private void fillRequest(MixPanelBatchSourceConfig config, HttpPost request) throws UnsupportedEncodingException {
+    List<NameValuePair> params = new LinkedList<>();
+    params.add(new BasicNameValuePair("from_date", config.getFromDate()));
+    params.add(new BasicNameValuePair("to_date", config.getToDate()));
+    String events = config.getEvents();
+    if (events != null && !events.isEmpty()) {
+      params.add(new BasicNameValuePair("event", events));
+    }
+    String filter = config.getFilter();
+    if (filter != null && !filter.isEmpty()) {
+      params.add(new BasicNameValuePair("where", filter));
+    }
+    request.setEntity(new UrlEncodedFormEntity(params));
+  }
+
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) throws IOException {
+    Configuration conf = taskAttemptContext.getConfiguration();
+    String configJson = conf.get(MixPanelInputFormatProvider.PROPERTY_CONFIG_JSON);
+    MixPanelBatchSourceConfig config = gson.fromJson(configJson, MixPanelBatchSourceConfig.class);
+
+    HttpClientContext context = createContext(config);
+
+    httpClient = HttpClients.createDefault();
+    HttpPost request = new HttpPost(config.getMixPanelUrl());
+    fillRequest(config, request);
+
+    response = httpClient.execute(request, context);
```

Comment:
does this mean whole data is held in the response? Is there hadoop connector for mixpanel integration?

Reply:
Data retrieved by scanner sequentially from response stream, it is not grabbed completely after response object returned.

There is no connectors for this. Api is very simple itself - single HTTP query returning single JSONL file(each line of file is some json object representing event).

Reply:
So basically response contents has all the events and scanner just scans through each line right? That means all the content will be held in memory?